cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -144,61 +146,64 @@ protected String processMessages(Iterator<ImmutablePair<String, T>> messages) th
                                                                                               timeBoundMillis);
       MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
       processMessages(context, trackingIterator);
-      String lastMessageId = trackingIterator.getLastMessageId();
-
-      // Persist the message id of the last message being consumed from the iterator
-      if (lastMessageId != null) {
-        storeMessageId(context, lastMessageId);
+      ImmutablePair<String, T> lastConsumer = trackingIterator.getLastComputedMessage();
+      if (lastConsumer != null) {
+        storeMessageId(context,  lastConsumer.getFirst());
       }
       return trackingIterator;
     }, Exception.class);
 
-    return iterator.getLastMessageId();
+    return iterator.getLastComputedMessage() != null ? iterator.getLastComputedMessage().getFirst() : null;
   }
 
   /**
    * An {@link Iterator} that remembers the message id that has been consumed up to.
    */
-  private final class MessageTrackingIterator extends AbstractIterator<ImmutablePair<String, T>> {
+  @VisibleForTesting
+  public final class MessageTrackingIterator implements Iterator<ImmutablePair<String, T>> {
 
-    private final Iterator<ImmutablePair<String, T>> messages;
-    private String lastMessageId;
+    private final PeekingIterator<ImmutablePair<String, T>> messages;
     private int consumedCount;
-    private boolean shouldEnd;
+    private ImmutablePair<String, T> lastComputedMessage;
 
-    MessageTrackingIterator(Iterator<ImmutablePair<String, T>> messages) {
+    public MessageTrackingIterator(PeekingIterator<ImmutablePair<String, T>> messages) {
       this.messages = messages;
       this.consumedCount = 0;
-      this.shouldEnd = false;
+      this.lastComputedMessage = null;
+    }
+
+    @Nullable
+    ImmutablePair<String, T>  getLastComputedMessage() {
```

Comment:
extra space in front of getLastComputedMessage

Reply:
Fixed